src/System.Console/src/System/ConsolePal.Unix.cs

```diff
@@ -170,6 +253,129 @@ public static bool CursorVisible
             }
         }
 
+        // TODO: It's quite expensive to use the request/response protocol each time CursorLeft/Top is accessed.
+        // We should be able to (mostly) track the position of the cursor in locals, doing the request/response infrequently.
+
+        public static int CursorLeft
+        {
+            get
+            {
+                int left, top;
+                GetCursorPosition(out left, out top);
+                return left;
+            }
+        }
+
+        public static int CursorTop
+        {
+            get
+            {
+                int left, top;
+                GetCursorPosition(out left, out top);
+                return top;
+            }
+        }
+
+        /// <summary>Gets the current cursor position.  This involves both writing to stdout and reading stdin.</summary>
+        private static unsafe void GetCursorPosition(out int left, out int top)
+        {
+            left = top = 0;
+
+            // Getting the cursor position involves both writing out a request string and
+            // parsing a response string from the terminal.  So if anything is redirected, bail.
+            if (Console.IsInputRedirected || Console.IsOutputRedirected)
+                return;
+
+            // Get the cursor position request format string.
+            string cpr = TerminalBasicInfo.Instance.CursorPositionRequestFormat;
+            if (string.IsNullOrEmpty(cpr))
+                return;
+
+            // Synchronize with all other stdin readers.  We need to do this in case multiple threads are
+            // trying to read/write concurrently, and to minimize the chances of resulting conflicts.
+            // This does mean that Console.get_CursorLeft/Top can't be used concurrently Console.Read*, etc.;
+            // attempting to do so will block one of them until the other completes, but in doing so we prevent
+            // one thread's get_CursorLeft/Top from providing input to the other's Console.Read*.
+            lock (StdInReader) 
+            {
+                // Write out the cursor position request.
+                WriteStdoutAnsiString(cpr);
+
+                // Read the response.  There's a race condition here if the user is typing,
+                // or if other threads are accessing the console; there's relatively little
+                // we can do about that, but we try not to lose any data.
+                StdInStreamReader r = StdInReader.Inner;
+                const int BufferSize = 1024;
+                byte* bytes = stackalloc byte[BufferSize];
+
+                int bytesRead = 0, i = 0;
+
+                // Response expected in the form "\ESC[row;colR".  However, user typing concurrently
```

Comment:
> the protocol involves writing to stdout, at which point data gets put on stdin for reading.

Yeah, it sounds like given the mechanism we have, this is the best we can do, other than tweaking things here and there.

> we could look into having some fields attempt to track the current cursor position, e.g. get the values once at the start of the app, and then based on the keys that are read, try to update those values to be kept in sync with the real cursor position

I was going to suggest something like this, as it seemed sort of similar to what we are now doing for ConsoleColor. On the other hand, it seems way riskier to do it for tracking the position because any tracking errors will accumulate, whereas the color gets "reset" every time you change it, so if you were out-of-sync with the previous color, it doesn't really matter now that you know the new color.